src/jllvm/gc/GarbageCollector.hpp

```diff
@@ -69,30 +64,30 @@ class GCUniqueRoot : public GCRootRef<T>
     GCUniqueRoot& operator=(const GCUniqueRoot&) = delete;
 
     /// Transfers ownership of the root from 'rhs' to this newly constructed root.
-    /// 'rhs' is left in an invalid state on which no methods but the destructor, 'data' and the move assignment
-    /// operator are valid.
-    GCUniqueRoot(GCUniqueRoot&& rhs) noexcept
-        : GCRootRef<T>(std::exchange(rhs.m_object, nullptr)), m_gc(std::exchange(rhs.m_gc, nullptr))
+    /// 'rhs' is an empty root afterwards.
+    GCUniqueRoot(GCUniqueRoot&& rhs) noexcept : Base(rhs.release()), m_gc(rhs.m_gc) {}
+
+    GCUniqueRoot& operator=(GCUniqueRoot&& rhs) noexcept
     {
+        // Reuse move constructor for move assignment.
+        this->~GCUniqueRoot();
+        return *new (this) GCUniqueRoot(std::move(rhs));
     }
 
-    /// Assignment version of the move constructor. See the move constructor.
-    GCUniqueRoot& operator=(GCUniqueRoot&& rhs) noexcept
+    /// Assignment from nullptr.
+    /// Equivalent to calling 'reset'.
+    GCUniqueRoot& operator=(std::nullptr_t) noexcept
```

Comment:
Should this not also be called `assign` to be consistent?

Reply:
Thought about this for a bit, but I think not calling it `assign` is actually more consistent. What assign does is change the object the root is referring to, while this method resets the root. 

That said, I think making the root "disappear" that easily is probably a bit of a footgun and non-obvious, so I opted to delete the operator. We should just use `reset` as that shows intend more clearly